File: src/scraper.js
import puppeteer from 'puppeteer';
import UserAgent from 'user-agents';
import { config } from './config.js';
import { RealEstateListing } from './models.js';
import { logger } from './utils/logger.js';
import { sleep, randomInt } from './utils/helpers.js';

export class StealthScraper {
  async setupBrowser() {
    const userAgent = new UserAgent();
    
    return await puppeteer.launch({
      headless: true,
      args: [
        '--disable-blink-features=AutomationControlled',
        '--disable-dev-shm-usage',
        `--user-agent=${userAgent.toString()}`
      ]
    });
  }

  async simulateHumanBehavior(page) {
    // Random scroll
    await page.evaluate(() => {
      window.scrollTo({
        top: Math.random() * document.body.scrollHeight,
        behavior: 'smooth'
      });
    });
    
    await sleep(randomInt(500, 1500));

    // Random mouse movements
    for (let i = 0; i < randomInt(2, 5); i++) {
      await page.mouse.move(
        randomInt(0, 800),
        randomInt(0, 600)
      );
      await sleep(randomInt(100, 300));
    }
  }

  async scrapeKvEe(criteria) {
    const browser = await this.setupBrowser();
    const listings = [];
    
    try {
      const page = await browser.newPage();
      
      // Set custom headers
      await page.setExtraHTTPHeaders({
        'Accept-Language': 'et-EE,et;q=0.9,en-US;q=0.8,en;q=0.7',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1'
      });

      const url = this.buildSearchUrl(criteria);
      await page.goto(url);
      await this.simulateHumanBehavior(page);

      while (true) {
        const pageListings = await this.extractPageListings(page);
        listings.push(...pageListings);
        
        const nextButton = await page.$('.pagination__next');
        if (!nextButton) break;
        
        await sleep(randomInt(config.scraping.minDelay, config.scraping.maxDelay));
        await nextButton.click();
        await page.waitForNavigation();
        await this.simulateHumanBehavior(page);
      }
    } finally {
      await browser.close();
    }
    
    return listings;
  }

  buildSearchUrl(criteria) {
    const params = new URLSearchParams({
      city: criteria.location,
      type: criteria.propertyType,
      price_min: criteria.minPrice,
      price_max: criteria.maxPrice
    });
    
    return `${config.urls.kvEeBase}/search?${params.toString()}`;
  }

  async extractPageListings(page) {
    const listings = [];
    const elements = await page.$$('.real-estate-item');
    
    for (const element of elements) {
      try {
        const listing = new RealEstateListing({
          source: 'kv.ee',
          listingUrl: await this.extractAttribute(element, 'href'),
          price: await this.extractPrice(element),
          address: await this.extractAddress(element),
          areaM2: await this.extractArea(element),
          rooms: await this.extractRooms(element),
          listingTime: new Date(),
          contactInfo: await this.extractContact(element)
        });
        
        listings.push(listing);
      } catch (error) {
        logger.error('Error extracting listing:', error);
      }
    }
    
    return listings;
  }

  // Helper methods to be implemented based on actual HTML structure
  async extractAttribute(element, attr) {
    return await element.evaluate((el, attr) => el.getAttribute(attr), attr);
  }

  async extractPrice(element) {
    // Implementation needed
    return 0;
  }

  async extractAddress(element) {
    // Implementation needed
    return '';
  }

  async extractArea(element) {
    // Implementation needed
    return 0;
  }

  async extractRooms(element) {
    // Implementation needed
    return null;
  }

  async extractContact(element) {
    // Implementation needed
    return null;
  }
}

File: src/scraper.py
from playwright.async_api import async_playwright, Browser, Page
from fake_useragent import UserAgent
import asyncio
import random
from datetime import datetime
from loguru import logger
from typing import List, Optional

from models import RealEstateListing
from config import SearchCriteria, Config

class StealthScraper:
    def __init__(self):
        self.ua = UserAgent()
        
    async def _setup_browser(self) -> Browser:
        """Setup a stealth browser instance with randomized properties."""
        playwright = await async_playwright().start()
        browser = await playwright.chromium.launch(
            headless=True,
            args=[
                '--disable-blink-features=AutomationControlled',
                '--disable-dev-shm-usage',
                f'--user-agent={self.ua.random}'
            ]
        )
        return browser

    async def _random_delay(self) -> None:
        """Implement randomized delay between actions."""
        delay = random.uniform(Config.MIN_DELAY, Config.MAX_DELAY)
        await asyncio.sleep(delay)

    async def _simulate_human_behavior(self, page: Page) -> None:
        """Simulate realistic human-like behavior."""
        # Random scroll
        await page.evaluate("""
            window.scrollTo({
                top: Math.random() * document.body.scrollHeight,
                behavior: 'smooth'
            });
        """)
        await self._random_delay()
        
        # Random mouse movements
        for _ in range(random.randint(2, 5)):
            await page.mouse.move(
                random.randint(0, 800),
                random.randint(0, 600)
            )
            await asyncio.sleep(random.uniform(0.1, 0.3))

    async def scrape_kv_ee(self, criteria: SearchCriteria) -> List[RealEstateListing]:
        """
        Scrape real estate listings from kv.ee with stealth measures.
        """
        listings = []
        browser = await self._setup_browser()
        
        try:
            page = await browser.new_page()
            
            # Set custom headers
            await page.set_extra_http_headers({
                'Accept-Language': 'et-EE,et;q=0.9,en-US;q=0.8,en;q=0.7',
                'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
                'DNT': '1',
                'Connection': 'keep-alive',
                'Upgrade-Insecure-Requests': '1',
            })

            # Navigate to search page
            search_url = self._build_search_url(criteria)
            await page.goto(search_url)
            await self._simulate_human_behavior(page)

            # Extract listings
            while True:
                # Extract current page listings
                page_listings = await self._extract_page_listings(page)
                listings.extend(page_listings)
                
                # Check for next page
                next_button = await page.query_selector('.pagination__next')
                if not next_button or not await next_button.is_visible():
                    break
                    
                await self._random_delay()
                await next_button.click()
                await page.wait_for_load_state('networkidle')
                await self._simulate_human_behavior(page)

        finally:
            await browser.close()
            
        return listings

    def _build_search_url(self, criteria: SearchCriteria) -> str:
        """Build the search URL based on criteria."""
        # Implementation specific to kv.ee URL structure
        base_url = f"{Config.KV_EE_BASE_URL}/search"
        params = {
            "city": criteria.location,
            "type": criteria.property_type.value,
            "price_min": criteria.min_price,
            "price_max": criteria.max_price,
        }
        # Convert params to URL query string
        return f"{base_url}?{'&'.join(f'{k}={v}' for k, v in params.items() if v is not None)}"

    async def _extract_page_listings(self, page: Page) -> List[RealEstateListing]:
        """Extract listings from the current page."""
        listings = []
        # Implementation specific to kv.ee DOM structure
        # This is a placeholder - actual implementation would need to match kv.ee's HTML structure
        listing_elements = await page.query_selector_all('.real-estate-item')
        
        for element in listing_elements:
            try:
                listing = RealEstateListing(
                    source="kv.ee",
                    listing_url=await element.get_attribute('href'),
                    price=await self._extract_price(element),
                    address=await self._extract_address(element),
                    area_m2=await self._extract_area(element),
                    rooms=await self._extract_rooms(element),
                    listing_time=datetime.utcnow(),  # Placeholder
                    contact_info=await self._extract_contact(element)
                )
                listings.append(listing)
            except Exception as e:
                logger.error(f"Error extracting listing: {e}")
                continue
                
        return listings

    # Helper methods for extracting specific fields
    async def _extract_price(self, element) -> float:
        # Implementation needed
        pass

    async def _extract_address(self, element) -> str:
        # Implementation needed
        pass

    async def _extract_area(self, element) -> float:
        # Implementation needed
        pass

    async def _extract_rooms(self, element) -> Optional[int]:
        # Implementation needed
        pass

    async def _extract_contact(self, element) -> Optional[str]:
        # Implementation needed
        pass